src/__demo__/Tooltip.stories.js

```diff
@@ -0,0 +1,94 @@
+import React from 'react'
+import { Tooltip } from '../Tooltip.js'
+import propTypes from '@dhis2/prop-types'
+
+export default {
+    title: 'Tooltip',
+    component: Tooltip,
+    decorators: [
+        storyFn => (
+            <div
+                style={{
+                    display: 'flex',
+                    alignItems: 'center',
+                    justifyContent: 'center',
+                    width: 500,
+                    height: 300,
+                }}
+            >
+                <div style={{ width: 300, height: 100 }}>{storyFn()}</div>
+            </div>
+        ),
+    ],
+}
+
+export const DefaultPlacementTop = () => (
+    <p>
+        I am a <Tooltip content="Some extra info">paragraph</Tooltip> that
+        contains a tooltip.
+    </p>
+)
+
+export const PlacementRight = () => (
+    <p>
+        I am a{' '}
+        <Tooltip placement="right" content="Some extra info">
+            paragraph
+        </Tooltip>{' '}
+        that contains a tooltip.
+    </p>
+)
+
+export const PlacementBottom = () => (
+    <p>
+        I am a{' '}
+        <Tooltip placement="bottom" content="Some extra info">
+            paragraph
+        </Tooltip>{' '}
+        that contains a tooltip.
+    </p>
+)
+
+export const PlacementLeft = () => (
+    <p>
+        I am a{' '}
+        <Tooltip placement="left" content="Some extra info">
+            paragraph
+        </Tooltip>{' '}
+        that contains a tooltip.
+    </p>
+)
+
+export const CustomElement = () => {
+    return (
+        <p>
+            I am a{' '}
+            <Tooltip content="Some extra info" component="p">
+                paragraph
+            </Tooltip>{' '}
+            that contains a tooltip.
+        </p>
+    )
+}
+
+const Content = React.forwardRef(({ children, ...rest }, ref) => (
+    <em style={{ display: 'inline-block', color: 'red' }} ref={ref} {...rest}>
+        {children}
+    </em>
+))
+Content.displayName = 'Content'
+Content.propTypes = {
+    children: propTypes.node,
+}
```

Comment:
I personally think this is quite involved / not really user-friendly. But it is very flexible ....

Reply:
Yeah, it moves the complexity from inside the component to the consumer.

But on the other hand, if you just need to swap the `span` for a `div`, this approach does work quite well since it is just `component="div"` that is required. The main criticism I have is that it is a bit obscure as it is not immediately obvious how the Tooltip wraps children internally.

It would be nice to handle this all transparently (as MUI does), but their Tooltip implementation is as usual more complex than what we need. They use `React.forwardRef(Tooltip)` and `React.cloneElement(children)` to accomplish it:

- https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/Tooltip/Tooltip.js#L178
- https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/Tooltip/Tooltip.js#L505

Reply:
I've looked at the MUI implementation today and there is currently something stopping us from implementing things that way. I'll explain:
In the MUI docs it says this about the `children` prop:
`children * | element |  Tooltip reference element.⚠️ *Needs to be able to hold a ref*.`
And in their examples they use a variety of MUI components that are wrapped directly in a MUI `Tooltip`. When inspecting these components you will see that *each one* is actually wrapped in `React.forwardRef`. That is the sole reason that line 505, where the element is being cloned is working correctly.

So if we want to go that way, which we might want to do, then one of the consequences would be that all of our components need to be able to hold a ref. That would be a pretty big change. Perhaps quite a good one. But in any case very substantial.

Reply:
Thinking about this some more, I'm not sure how we'd work with these event handlers either... I don't think we'd want every component to have implemented `onMouseOver` and `onMouseOut` too, on top of being able to hold a ref... I suppose we could just use the ref and add some native event listeners to that... But it'd be better to use React's synthetic events....

Quite an interesting problem 🤔

Reply:
The most basic approach would just be to wrap the children in some element (or perhaps allow switching between `span`, `p`, etc.), and allow some form of styling to make sure that element behaves properly. Shame this approach would be very inconsistent with the rest of the codebase.

Reply:
Would it be easier to build the Tooltip using the `react-popper` library?

https://github.com/popperjs/react-popper

Maybe we found a reason to use react-popper over popper.

Reply:
```jsx
    <Reference>
      {({ ref }) => (
        <button type="button" ref={ref}>
          Reference element
        </button>
      )}
    </Reference>
```
I don't think we should switch to `react-popper` (motivation available upon request), but perhaps we could use this pattern as a source of inspiration... Maybe we can implement a `TooltipReference` component that also does renderProps.

I think the most important thing to realise is that even if we were to implement this helper Component, or even switch to `react-popper`, the core problem remains:
- The popper needs a ref to a reference-element
- So whatever is being used as a Tooltip child needs to be able to hold a `ref`

So, thinking of a `Button` (ui-core component) with a `Tooltip` the following example would NOT work:
```jsx
    <Reference>
      {({ ref }) => (
        <Button ref={ref}>
          Reference element
        </Button>
      )}
    </Reference>
```

The reason is that our `Button` can't hold a ref. So the app would either need to implement a wrapper around the `Button` that is able hold a ref, or ui-core needs to make sure all/most of its components can hold a ref.

I'm not really in favour of reworking all of our component, just to make the Tooltip work[*], so letting the app wrap whatever they need seems to make more sense.

[*] Perhaps letting all of our components accept a ref isn't such a bad idea in general, but we shouldn't start adding this just because we need to fix the Tooltip.